pkg/certificateprovider/azurekeyvault/provider.go

```diff
@@ -0,0 +1,368 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package azurekeyvault
+
+// This class is based on implementation from  azure secret store csi provider
+// Source: https://github.com/Azure/secrets-store-csi-driver-provider-azure/blob/master/pkg/provider/
+import (
+	"context"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"path/filepath"
+	"reflect"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	kv "github.com/Azure/azure-sdk-for-go/services/keyvault/v7.1/keyvault"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/Azure/go-autorest/autorest/date"
+	"github.com/deislabs/ratify/pkg/certificateprovider/azurekeyvault/types"
+	"github.com/sirupsen/logrus"
+
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+)
+
+type keyvaultObject struct {
+	content string
+	version string
+}
+
+func GetCertificatesContent(ctx context.Context, attrib map[string]string) ([]types.CertificateFile, error) {
+
+	keyvaultName := types.GetKeyVaultName(attrib)
+	cloudName := types.GetCloudName(attrib)
+	tenantID := types.GetTenantID(attrib)
+	workloadIdentityClientID := types.GetClientID(attrib)
+
+	if keyvaultName == "" {
+		return nil, fmt.Errorf("keyvaultName is not set")
+	}
+	if tenantID == "" {
+		return nil, fmt.Errorf("tenantID is not set")
+	}
+	if workloadIdentityClientID == "" {
+		return nil, fmt.Errorf("clientId is not set")
+	}
+
+	azureCloudEnv, err := ParseAzureEnvironment(cloudName)
+	if err != nil {
+		return nil, fmt.Errorf("cloudName %s is not valid, error: %w", cloudName, err)
+	}
+
+	// 1. cleaning up keyvault objects definition
+	objectsStrings := types.GetCertificates(attrib)
+	if objectsStrings == "" {
+		return nil, fmt.Errorf("objects is not set")
+	}
+	logrus.Infof("objects string defined in secret provider class, objects %v", objectsStrings)
+
+	objects, err := types.GetCertificatesArray(objectsStrings)
+	if err != nil {
+		return nil, fmt.Errorf("failed to yaml unmarshal objects, error: %w", err)
+	}
+	logrus.Infof("unmarshaled objects yaml, objectsArray %v", objects.Array)
+
+	keyVaultCerts := []types.KeyVaultCertificate{}
+	for i, object := range objects.Array {
+		var keyVaultCert types.KeyVaultCertificate
+		err = yaml.Unmarshal([]byte(object), &keyVaultCert)
+		if err != nil {
+			return nil, fmt.Errorf("unmarshal failed for keyVaultCerts at index %d, error: %w", i, err)
+		}
+		// remove whitespace from all fields in keyVaultCert
+		formatKeyVaultCertificate(&keyVaultCert)
+		if err = validate(keyVaultCert); err != nil {
+			return nil, wrapObjectTypeError(err, keyVaultCert.CertificateName, keyVaultCert.CertificateVersion)
+		}
+
+		keyVaultCerts = append(keyVaultCerts, keyVaultCert)
+	}
+
+	logrus.Infof("unmarshaled key vault objects, keyVaultObjects %v , count %v", keyVaultCerts, len(keyVaultCerts))
+
+	if len(keyVaultCerts) == 0 {
+		return nil, nil
+	}
```

Comment:
If the keyvault is provided, is it required to provide the certs too for ratify? Should this be an error instead?